### ISSUE TITLE
clarify MPI behavior when multiple MPI processes run in the same address space

### ISSUE THREAD

**Issue by mpiforumbot** [mpi-forum/mpi-forum-historic#310]
**Originally by _snir_ on 2011-12-12 12:22:28 -0600**

---

The proposal clarifies how MPI behaves when multiple processes share the same address space. It adds a new spli_type constant to identify all MPI processes within the same address space.

A minimal implementation is trivial: It requires that MPI_COMM_SPLIT_TYPE recognize a new spli_type value and returns a duplicate of MPI_COMM_SELF 

The proposal depends on Ticket #287

**Comment by mpiforumbot** [comment]
**Originally by _snir_ on 2011-12-12 12:23:13 -0600**

---

**Attachment added:** [`main.pdf`](https://github.com/mpiforumbot/mpi-forum-historic/raw/master/attachments/310/main.pdf) (157.5 KiB)

**Comment by mpiforumbot** [comment]
**Originally by _dougmill_ on 2011-12-15 13:55:32 -0600**

---

Added an important set of comments to ticket #311, which probably should have been posted here instead.

**Comment by mpiforumbot** [comment]
**Originally by _balaji_ on 2012-01-01 10:25:47 -0600**

---

**Attachment added:** [`mpi-report.pdf`](https://github.com/mpiforumbot/mpi-forum-historic/raw/master/attachments/310/mpi-report.pdf) (3661.6 KiB)
Document for the first formal reading

**Comment by mpiforumbot** [comment]
**Originally by _balaji_ on 2012-01-01 10:26:54 -0600**

---

Attached a new document intended for the upcoming formal reading.  The document contains changes for this ticket and ticket #311.

**Comment by mpiforumbot** [comment]
**Originally by _balaji_ on 2012-01-01 10:30:12 -0600**

---

Dropping the entire hybrid mailing list from the cc list for this ticket.  Folks interested in following this particular ticket can add themselves to the cc list.

**Comment by mpiforumbot** [comment]
**Originally by _dougmill_ on 2012-01-03 08:19:23 -0600**

---

I'm still having problems with the declaration about inheritence. The sentence:

```
For example, it can store, in a thread-private location, a pointer to
the data structure associated with this MPI process that is inherited
from parent to child.
```

Doesn't really help. Since thread-private storage is not inherited by the child, the child thread still has no way to know what the parent had done, and when. Even if the child could somehow resolve a structure address from the parent, and make a new copy of it, the copy would be out of date and would not reflect conditions of the parent at the time the parent called (for example) pthread_create().

**Comment by mpiforumbot** [comment]
**Originally by _balaji_ on 2012-01-03 10:16:19 -0600**

---

The sentence is not meant to say it would automatically inherit this information.  All it is meant to say is that if an MPI process spawns a new thread, the new thread would have the same rank (e.g., by copying the rank information).  This can be done, for example, by sticking this information into a location on the heap indexed by the thread ID, which all threads can access.  This is already true today, not something that is being proposed.  The sentence is only meant as a clarification.  If the text is not clear enough, can you suggest a replacement?

With respect to this information becoming stale, I don't understand this at all -- this information is constant as far as this ticket is concerned.

**Comment by mpiforumbot** [comment]
**Originally by _dougmill_ on 2012-01-03 10:35:34 -0600**

---

The problem, as I see it, is that the child thread has no way of determining where the parent thread would have stashed that information. A child thread cannot, at least in the pthread world, determine who the parent was, let alone any sort of ordinal thread ID to index into a table.

If the child does have a way to access the parent's "current attached endpoint" then that information is out of date by the time the child goes to copy it as it's "initial attached endpoint". The parent very likely could call MPI_THREAD_ATTACH after spawning thread(s), which means the attached endpoint(s) that the child thread(s) get might not even be consistent, let alone might not reflect what the parent was attached to at the time it called pthread_create.

I don't believe the semantics being describe (inheritance of attched endpoint) can be implemented, or at least cannot be implemented in a manner that has predictable outcome. I don't think it will be acceptible to require MPI+thread programmers to perform special operations before and/or after calling pthread_create (or OpenMP blocks), in order to implement the inheritance. I think the issue of what endpoint a new thread is attached to is important and must be documented, but I think the only practical semantics are that new threads are attached to "endpoint 0".

**Comment by mpiforumbot** [comment]
**Originally by _balaji_ on 2012-01-03 10:46:45 -0600**

---

I think you are confusing between many things.
1. What is an endpoint?  The document has no mention of an endpoint.
2. What is thread attach?  This ticket does not deal with thread migration.  (I agree this is an issue with thread migration, but please keep that discussion in the correct ticket).

With respect to not knowing my parent, I understand your concern and I'll look into what MPI implementations like MPC are doing for this.

**Comment by mpiforumbot** [comment]
**Originally by _dougmill_ on 2012-01-03 11:02:50 -0600**

---

endpoint ==> "MPI Process", although it's not entirely clear to me if your current definition of MPI Process represents the same thing that "endpoints" used to, or at least what I took that to mean.

Unfortunately, while the issue may only arise when MPI_THREAD_ATTACH is present but the semantics which create the problem are described under this ticket.

I realize that the text I talking about is under "Advice to implementers". Perhaps it shouldn't be mentioned at all. But, an obvious question when one reads this is "what am I attached to when I start". I think the answer should be in the text, probably not under "Advice", and should be possible to implement.

Is it your intent to leave the behavior "undefined"? That essentially requires everyone to make a MPI_THREAD_ATTACH call before doing any MPI operations in a new thread.  Perhaps that is OK, but it needs to be clearly stated, and not under any sort of "Advice" header.

**Comment by mpiforumbot** [comment]
**Originally by _balaji_ on 2012-01-03 11:12:50 -0600**

---

The concept of "endpoints" was creating too much confusion, and hence was dropped.  I'm not sure what your understanding was, but we didn't lose any semantic value from the initial discussion on endpoints.

Once again, this ticket does not mean to propose any new semantics.  These semantics were always allowed in MPI (the text already says MPI processes can be different address spaces or in a shared address space).  This ticket only clarifies this.  The advice to implementers is also just clarification of something that was always allowed.  So dropping it is not something that will fix this problem.  What you are pointing out is a potential problem that has always existed in MPI (even MPI-2.0), and not something we are introducing.  The only thing that is left to be figured out, is if it is in fact a problem.

And no, requiring everyone to make an MPI_THREAD_ATTACH call is not OK.  The moment we talk about THREAD_ATTACH, this ticket will no longer be a "clarification only" ticket.  Please keep that discussion on ticket #311, where it belongs.

**Comment by mpiforumbot** [comment]
**Originally by _dougmill_ on 2012-01-03 12:21:55 -0600**

---

I think I'll be OK with it if the phrase "that is inherited from parent to child" is removed. Again, this is the paraphraph on page 22, section 2.7, which is marked with "ticket310" (this ticket).

**Comment by mpiforumbot** [comment]
**Originally by _jsquyres_ on 2012-01-05 13:40:23 -0600**

---

I have a knee-jerk reaction against adding the "-asp" parameter to mpiexec (and corresponding info key to COMM_SPAWN).  I know that all parameters to mpiexec (and info keys to COMM_SPAWN) are optional, but we do have language to the effect of "if you have mpiexec, we encourage you to support all of these parameters..."  

Again, I know that these are optional, and even if, for example, mpirun has an -asp option, it could certainly fail for all cases where the asp value was >1.  But that's still... icky.

In effect, this ticket strongly encouraging all MPI implementations to support the multi-MPI-process-per-address-space model.  Yow!  Adding an mpiexec option (COMM_SPAWN info key) is a weird way to do that.  If you want MPI to support this model, why not be more explicit about it?

**Comment by mpiforumbot** [comment]
**Originally by _snir_ on 2012-01-05 16:52:50 -0600**

---

The proposal is not strongly encouraging all implementations ot support multi-MPI_process-per-address_space, and is not weakly encouraging it. It is just providing a standard way of doing it for implementers that are inclined to do so (and there are several such). Please do not read hidden agendas in the proposal: I may personally believe that such support is a good idea, but I do not try to convince the MPI standard committee of such -- merely that it is a good idea to standardize how this is done, since the impact on the standard is ridiculously slim.  If the idea is atrocious, then nobody will ask for it and nobody will implement it, and there will be no change what so ever in user codes or MPI implementations. If the idea is good, then there is a way to support it in a standard manner; users will ask for this support and implementers will supply it. 

Again, I do not want MPI to support a particular model -- this is already a given: Some MPI implementations do run multiple MPI processes within one address space. I want to standardize how such a computation is started -- to the extent MPI initialization is standardized (which is limited). 

Incidentally, the reason MPI1 does explicitly mention the possibility of having multiple MPI processes within one address space is because one of the earliest implementation of MPI (for NEC) did so.

**Comment by mpiforumbot** [comment]
**Originally by _jsquyres_ on 2012-01-06 12:15:25 -0600**

---

Minor note: It should be noted that this ticket is dependent on ticket #287, which is getting its 2nd vote at the January 2012 meeting.

**Comment by mpiforumbot** [comment]
**Originally by _jsquyres_ on 2012-01-06 12:21:25 -0600**

---

I just noticed that the PDF on this ticket (#310) is missing the table added in #287 entitled "Communicator split type constants" in appendix A (page 540 in the fulldoc-v0.6.pdf on #287).  I'm guessing this was a copy-n-paste error from the #287 ticket.

I mention this because the new constant MPI_COMM_TYPE_ADDRESS_SPACE should be added to the table that was removed.

**Comment by mpiforumbot** [comment]
**Originally by _snir_ on 2012-01-06 13:34:33 -0600**

---

Thanks -- will add.

Please look at the mpi-report.pdf file if you want to comment on 310/311. That's the version to be voted on,

**Comment by mpiforumbot** [comment]
**Originally by _snir_ on 2012-01-08 17:13:38 -0600**

---

**Attachment added:** [`mpi-report1.pdf`](https://github.com/mpiforumbot/mpi-forum-historic/raw/master/attachments/310/mpi-report1.pdf) (2259.2 KiB)
added missing const table and MPI_INIT comment

**Comment by mpiforumbot** [comment]
**Originally by _snir_ on 2012-01-09 17:41:36 -0600**

---

**Attachment added:** [`mpi-report2.pdf`](https://github.com/mpiforumbot/mpi-forum-historic/raw/master/attachments/310/mpi-report2.pdf) (2281.8 KiB)
slight cleanup

**Comment by mpiforumbot** [comment]
**Originally by _dougmill_ on 2012-02-01 07:41:47 -0600**

---

I am concerned about something that was brought up in San Jose, the scope of handles. If the scope of handles must be per-MPI process then I think we have a scaling issue. The standard actually admits that it is oriented towards 1 MPI process per OS process but just infers that "other implementations are possible". I am hoping there is not an actual requirement that handles cannot be shared amongst threads in an OS process will have significant impacts. Imagine 64 threads/MPI processes per OS process then requiring 64x the number of handles.  This is clearly sub-optimal, especially when (most?) many implementations scope the handles to the OS process already.

**Comment by mpiforumbot** [comment]
**Originally by _jsquyres_ on 2012-02-01 08:09:54 -0600**

---

Replying to dougmill:

> I am concerned about something that was brought up in San Jose, the scope of handles. If the scope of handles must be per-MPI process then I think we have a scaling issue. The standard actually admits that it is oriented towards 1 MPI process per OS process but just infers that "other implementations are possible". 

I think it is fine for multiple MPI processes within the same OS process to share handle values if the MPI implementation chooses to do so.  For example, the value of the MPI_INT handle in MPI_COMM_WORLD rank 0 (which is thread A) may be identical to the value of the MPI_INT handle in MPI_COMM_WORLD rank 1 (which is thread B in the same OS process as MCW rank 0).  They may even both refer to the same underlying MPI object.

It gets trickier with MPI objects that need per-MPI-process state (e.g., communicators), but I suppose it can still be done.

I think the issue that was raised in San Jose was a derivative of the global variables issue in the multiple-MPI-processes-in-a-single-OS-process model.  I.e., if one MPI process (i.e., thread) happens to fill the MPI_Comm handle "foo", and "foo" happens to be a global variable, another MPI process (i.e., thread in the same OS process) cannot rely on that value.  It needs to have its own handle variable for that same communicator, for example.  The two handle variables may actually have the same value and point to the same underlying MPI object, but they need to be in different _variables_.

Right?

**Comment by mpiforumbot** [comment]
**Originally by _dougmill_ on 2012-02-01 08:24:52 -0600**

---

OK, if it's only an issue of needing multiple (thread private) variables for handles (that might vary between threads), that is alright. As long as we don't have to (always) create multiple, identical, handles.

**Comment by mpiforumbot** [comment]
**Originally by _jsquyres_ on 2012-02-01 08:37:04 -0600**

---

Replying to dougmill:

> OK, if it's only an issue of needing multiple (thread private) variables for handles (that might vary between threads), that is alright. As long as we don't have to (always) create multiple, identical, handles.

Did you mean handles, or objects?

My point is that an MPI implementation that uses the multiple-MPI-processes-per-OS-process model can have a single MPI object that has multiple handles of the same value in different MPI processes in the same OS process.  There's two cases:
- predefined handles (like MPI_INT): there's obviously a single variable for these in the OS process address space, and it needs to be valid in all MPI processes in that OS process
- user-created handles (e.g., the output of MPI_COMM_DUP): each thread will need to have their own variable for holding its own handle value.

Another example: if an OS process has 4 threads, each of which are MPI processes, if they all call MPI_Comm_dup(MPI_COMM_WORLD, &my_thread_local_comm_variable), then there are 4 MPI_Comm variables: one for each thread.  If the MPI implementation wants to, they can all point to the same underlying MPI object (i.e., 1 underlying object instead of 4).

But regardless -- MPI handles have to be treated just like any other global variables.  Their values cannot be shared between MPI processes via shared memory (e.g., global variables).  It is also erroneous to MPI_Send a handle value from one process to another -- while it may _work_ for MPI processes in the same OS process, it is definitely not _portable_.

**Comment by mpiforumbot** [comment]
**Originally by _dougmill_ on 2012-02-01 08:43:39 -0600**

---

My point is that threads != MPI-processes, so a program might create MPI_COMM_CLIQUE as the set of local "MPI processes" extracted from MPI_COMM_WORLD.  All threads, no matter what MPI process they are currently attached to, should be able to use that single handle for MPI_COMM_CLIQUE.

**Comment by mpiforumbot** [comment]
**Originally by _jsquyres_ on 2012-02-01 13:28:38 -0600**

---

I don't know what MPI_COMM_CLIQUE is.

I don't think I agree with your statement: "All threads, no matter what MPI process they are currently attached to, should be able to use that single handle for MPI_COMM_CLIQUE" -- unless MPI_COMM_CLIQUE is a pre-defined MPI handle...?

If it's not predefined, you can't guarantee that the handle value will be the same in all MPI processes.  It's up to the implementation.

**Comment by mpiforumbot** [comment]
**Originally by _dougmill_ on 2012-02-03 07:50:28 -0600**

---

The standard (2.2) does not differentiate between predefined and user-created handles in this respect, it says only:

```
An opaque object and its handle are significant only
at the process where the object was created and cannot
be transferred to another process.
```

This creates a problem for predefined handles with multiple MPI processes, e.g. where can MPI_COMM_WORLD be used?  But I am less interested in what the standard has said in the past, and more what we can say going forward.  I think this issue will be significant to "scaling inwards" (towards threads and multiple MPI-processes).

That being said, there are oddities with allowing handles to escape the MPI-process. Right now, at least in most implementations (1:1 OS-process:MPI-process), a piece of code cannot ever get an error from using MPI_Comm_rank (for example) on a communicator that it knows about. With multiple MPI-processes per OS-process, a given thread will be "attached" to only one MPI-process and that MPI-process may not be in all communicators created in that OS-process address space (visible to the threads).  This also makes MPI_COMM_SELF a very strange object, as it would have to have different values depending on the threads that call it, and what they are attached to when they call. Then there is the fact that the value returned by `MPI_Comm_rank(MPI_COMM_WORLD, &rank)` might change throughout the life of a program.

How this is defined will have a profound impact on implementations and how efficient they can be. Having to put everything in thread-local-storage, and use other per-thread mechanisms, is going to negatively impact performance.

**Comment by mpiforumbot** [comment]
**Originally by _snir_ on 2012-02-22 08:53:08 -0600**

---

**Attachment added:** [`ticket 310-311.pdf`](https://github.com/mpiforumbot/mpi-forum-historic/raw/master/attachments/310/ticket%20310-311.pdf) (293.0 KiB)

**Comment by mpiforumbot** [comment]
**Originally by _snir_ on 2012-03-04 23:54:02 -0600**

---

**Attachment added:** [`ticket 310.pdf`](https://github.com/mpiforumbot/mpi-forum-historic/raw/master/attachments/310/ticket%20310.pdf) (331.0 KiB)
separated 310 and 311

**Comment by mpiforumbot** [comment]
**Originally by _snir_ on 2012-03-07 02:37:46 -0600**

---

**Attachment added:** [`ticket 310 v2.pdf`](https://github.com/mpiforumbot/mpi-forum-historic/raw/master/attachments/310/ticket%20310%20v2.pdf) (322.7 KiB)

**Comment by mpiforumbot** [comment]
**Originally by _jsquyres_ on 2012-06-20 09:53:01 -0500**

---

1st vote failed in Japan Forum meeting, May, 2012.  Moved to "author rework".
